core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -750,21 +750,20 @@ private[log] class Cleaner(val id: Int,
                                  batch: RecordBatch,
                                  record: Record,
                                  stats: CleanerStats): Boolean = {
+    // keep all records that are beyond the latest offset based on OffsetMap
     val pastLatestOffset = record.offset > map.latestOffset
     if (pastLatestOffset)
       return true
 
     if (record.hasKey) {
-      val key = record.key
-      val foundOffset = map.get(key)
-      /* First,the message must have the latest offset for the key
+      /* First,the message must have the latest offset or version for the key
        * then there are two cases in which we can retain a message:
        *   1) The message has value
        *   2) The message doesn't has value but it can't be deleted now.
        */
-      val latestOffsetForKey = record.offset() >= foundOffset
+      val isLatestRecordForKey = map.shouldRetainRecord(record)
```

Comment:
suggestion : call  map.shouldRetainRecord(record) after  checking for record.hasValue || retainDeletes

Reply:
its basically calculates both and make the decision later... so no impact in doing first or next.